freqtrade/freqai/data_kitchen.py

```diff
@@ -0,0 +1,1112 @@
+        # model load from disk
+        self.data["training_features_list_raw"] = copy.deepcopy(self.training_features_list)
+        self.training_features_list = self.data_dictionary["train_features"].columns
+
+        if self.freqai_config.get('data_split_parameters', {}).get('test_size', 0.1) != 0:
+            self.data_dictionary["test_features"] = pd.DataFrame(
+                data=test_components,
+                columns=["PC" + str(i) for i in range(0, n_keep_components)],
+                index=self.data_dictionary["test_features"].index,
+            )
+
+        self.data["n_kept_components"] = n_keep_components
+        self.pca = pca2
+
+        logger.info(f"PCA reduced total features from  {n_components} to {n_keep_components}")
+
+        if not self.data_path.is_dir():
+            self.data_path.mkdir(parents=True, exist_ok=True)
+
+        return None
+
+    def pca_transform(self, filtered_dataframe: DataFrame) -> None:
+        """
+        Use an existing pca transform to transform data into components
+        :params:
+        filtered_dataframe: DataFrame = the cleaned dataframe
+        """
+        pca_components = self.pca.transform(filtered_dataframe)
+        self.data_dictionary["prediction_features"] = pd.DataFrame(
+            data=pca_components,
+            columns=["PC" + str(i) for i in range(0, self.data["n_kept_components"])],
+            index=filtered_dataframe.index,
+        )
+
+    def compute_distances(self) -> float:
+        """
+        Compute distances between each training point and every other training
+        point. This metric defines the neighborhood of trained data and is used
+        for prediction confidence in the Dissimilarity Index
+        """
+        logger.info("computing average mean distance for all training points")
+        tc = self.freqai_config.get("model_training_parameters", {}).get("thread_count", -1)
+        pairwise = pairwise_distances(self.data_dictionary["train_features"], n_jobs=tc)
+        avg_mean_dist = pairwise.mean(axis=1).mean()
+        logger.info(f"avg_mean_dist {avg_mean_dist:.2f}")
+
+        return avg_mean_dist
+
+    def use_SVM_to_remove_outliers(self, predict: bool) -> None:
+        """
+        Build/inference a Support Vector Machine to detect outliers
+        in training data and prediction
+        :params:
+        predict: bool = If true, inference an existing SVM model, else construct one
+        """
+
+        if self.keras:
+            logger.warning(
+                "SVM outlier removal not currently supported for Keras based models. "
+                "Skipping user requested function."
+            )
+            if predict:
+                self.do_predict = np.ones(len(self.data_dictionary["prediction_features"]))
+            return
+
+        if predict:
+            assert self.svm_model, "No svm model available for outlier removal"
```

Comment:
Please don't assert in "non-test" code - asserts can be deactivated via a CLI flag of python ... so you can't rely on this to work on all cases.